Source/Protected/ProtectedAsMock.cs

```diff
@@ -0,0 +1,231 @@
+using System.Linq.Expressions;
+using System.Reflection;
+
+using Moq.Language.Flow;
+
+namespace Moq.Protected
+{
+	internal sealed class ProtectedAsMock<T, TDuck> : IProtectedAsMock<T, TDuck>
+		where T : class
+		where TDuck : class
+    {
+		private Mock<T> mock;
+
+		private static DuckReplacer DuckReplacerInstance = new DuckReplacer(typeof(TDuck), typeof(T));
+
+		public ProtectedAsMock(Mock<T> mock)
+		{
+			this.mock = mock;
+		}
+
+		public ISetup<T> Setup(Expression<Action<TDuck>> expression)
+		{
+			return Mock.Setup(this.mock, (Expression<Action<T>>)ReplaceDuck(expression), null);
+		}
+
+		public ISetup<T, TResult> Setup<TResult>(Expression<Func<TDuck, TResult>> expression)
+		{
+			return Mock.Setup(this.mock, (Expression<Func<T, TResult>>)ReplaceDuck(expression), null);
+		}
+
+		private static LambdaExpression ReplaceDuck(LambdaExpression expression)
+		{
+			var targetParameter = Expression.Parameter(typeof(T), expression.Parameters[0].Name);
+			return Expression.Lambda(DuckReplacerInstance.Visit(expression.Body), targetParameter);
+		}
+
+		/// <summary>
+		/// <see cref="ExpressionVisitor"/> used to replace occurrences of `TDuck.Member` sub-expressions with `T.Member`.
+		/// </summary>
+		private sealed class DuckReplacer : ExpressionVisitor
+		{
+			private Type duckType;
+			private Type targetType;
+
+			public DuckReplacer(Type duckType, Type targetType)
+			{
+				this.duckType = duckType;
+				this.targetType = targetType;
+			}
+
+			protected override Expression VisitMethodCall(MethodCallExpression node)
+			{
+				if (node.Object is ParameterExpression left && left.Type == this.duckType)
+				{
+					var targetParameter = Expression.Parameter(typeof(T), left.Name);
+					return Expression.Call(targetParameter, FindCorrespondingMethod(node.Method), node.Arguments);
+				}
+				else
+				{
+					return node;
+				}
+			}
+
+			protected override Expression VisitMember(MemberExpression node)
+			{
+				if (node.Expression is ParameterExpression left && left.Type == this.duckType)
+				{
+					var targetParameter = Expression.Parameter(typeof(T), left.Name);
+					return Expression.MakeMemberAccess(targetParameter, FindCorrespondingMember(node.Member));
+				}
+				else
+				{
+					return node;
+				}
+			}
+
+			private MemberInfo FindCorrespondingMember(MemberInfo duckMember)
+			{
+				if (duckMember is MethodInfo duckMethod)
+				{
+					return FindCorrespondingMethod(duckMethod);
+				}
+				else if (duckMember is PropertyInfo duckProperty)
+				{
+					return FindCorrespondingProperty(duckProperty);
+				}
+				else
+				{
+					throw new NotSupportedException();
+				}
+			}
+
+			private MethodInfo FindCorrespondingMethod(MethodInfo duckMethod)
+			{
+				var candidateTargetMethods =
+					this.targetType
+				    .GetMethods(BindingFlags.NonPublic | BindingFlags.Instance)
+				    .Where(ctm => IsCorrespondingMethod(duckMethod, ctm));
+
+				var targetMethod = candidateTargetMethods.Single();
```

Comment:
Should there be a check whether exactly one target method was identified (and if not, throw an exception with a meaningful `Message`)?